e2e/cluster/cluster.go

```diff
@@ -0,0 +1,503 @@
+package cluster
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"net"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/Azure/agentbakere2e/config"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/to"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/containerservice/armcontainerservice"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/network/armnetwork"
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/yaml"
+)
+
+// WARNING: if you modify cluster configuration, please change the version below
+// this will avoid potential conflicts with tests running on other branches
+// there is no strict rules or a hidden meaning for the version
+// testClusterNamePrefix is also used for versioning cluster configurations
+const testClusterNamePrefix = "abe2e-v20240704-"
+
+var (
+	clusterKubenet       *Cluster
+	clusterKubenetAirgap *Cluster
+	clusterAzureNetwork  *Cluster
+
+	clusterKubenetError       error
+	clusterKubenetAirgapError error
+	clusterAzureNetworkError  error
+
+	clusterKubenetOnce       sync.Once
+	clusterKubenetAirgapOnce sync.Once
+	clusterAzureNetworkOnce  sync.Once
+)
+
+type Cluster struct {
+	Model *armcontainerservice.ManagedCluster
+	Kube  *Kubeclient
+}
+
+type Kubeclient struct {
+	Dynamic client.Client
+	Typed   kubernetes.Interface
+	Rest    *rest.Config
+}
+
+// Returns true if the cluster is configured with Azure CNI
+func (c *Cluster) IsAzureCNI() (bool, error) {
+	if c.Model.Properties.NetworkProfile != nil {
```

Comment:
feels kinda clunky to check for `nil`ness in so many different places...

do you think adding a single place where `nil`ness validation takes place, immediately proceeding the relevant azure calls?

Reply:
This function called during each VMSS creation. I don't know details, but it seems like VMSS needs to be configured differently based on the flag.

Reply:
don't worry about it for now, I mainly just meant that it would be nice if we could somehow consolidate where we check for `nil` on certain objects